[AIRFLOW-2840] - add update connections cli option

caddac opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-2840

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
Adds an Update Connection CLI option. This is useful for updating connections in a CI/CD environment where deleting and adding connections could cause DAGS to fail.

### Tests

- [x] My PR adds the following unit tests:
My PR adds a `update connection` step to an existing unit test in `test/core.py`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

ashb reviewed (changes requested):
connection ID is (by design) non-unique - its a mechanism of load balancing across multiple DBs for instance - you can have three conns with an ID of "mysql" and Airflow will pick a random one each time.

Can you explicitly handle/detect this case (probably by refusing to update any of them?) and add tests that cover it too? Given it's relatively infrequent to do this I think it is okay to just refuse to update any of the conns in this case.

ashb commented on airflow/api/client/local_client.py:
```diff
@@ -51,3 +51,83 @@ def create_pool(self, name, slots, description):
+        :param conn_extra:
+        :return: The new Connection
+        """
+        return connections.add_connection(conn_id,
```
We probably shouldn't return objects over the API  (as only the "Local" api can do this - we couldn't return an object over JSON for instance) - this should be the conn_id that we return I think.

It is possibly worth returning the numeric ID (the `id` column, not to be confused with the `conn_id` column!) as this would let us delete individual connections if there are multiple ones with the same conn_id.

Similarly for the other APIs- ids or JSON representations of the conn.

caddac replied:
So the idea here was the client(s) would return the same objects (regardless of client type) and leave formatting the response to the api (cli or http). What about returning the json representation from the client, (which the local and json clients could do) and still leaving formatting to the API. TBH, the differences in clients confuses me and I'm still not sure what  or why we have two different clients given that the end users seem to interact with the API rather than the client directly. Any clarity you could provide there would be appreciated.

ashb reviewed (commented):
Haven't had time to give this a full re-review yet, sorry.

caddac committed:
[AIRFLOW-2840] - add update connections cli option

Adds an Update Connection CLI option.\nThis is useful for updating connections in a CI/CD environment\n where deleting and adding connections\n could cause DAGS to fail. \nAlso exposes connection api via rest

ashb commented on airflow/bin/cli.py:
```diff
@@ -1110,133 +1108,77 @@ def version(args):
-            return
+        conns = api_client.list_connections()
+        # format it for the cli
+        conns = list(map(lambda c: (c['conn_id'], c['conn_type'],
```
Minor point: we don't need the `list()` here

ashb commented on airflow/bin/cli.py:
```diff
@@ -1110,133 +1108,77 @@ def version(args):
+        try:
+            print(api_client.delete_connection(args.conn_id, args.delete_all))
+        except MissingArgument as ma:
+            print(ma)
```
We should exit with a Non-zero code here.

caddac replied:
is raising the exception good enough? or is there a specific way to exit with non-zero code?

ashb replied:
Raising the `ma` exception would work but would provide a messy output (i.e. it would include the stack trace which is not useful here):

https://github.com/apache/airflow/blob/9f3eae5e82aef4550027e5c3163715dad67bdb57/airflow/bin/cli.py#L1295-L1296

```
        except MissingArgument as ma:
            raise SystemExit(ma)
```

Should probably do the trick (and not include any errant stack-traces. If that still includes a stack trace, or errors then `SystemExit(str(ma)))`

ashb commented on airflow/bin/cli.py:
```diff
@@ -1110,133 +1108,77 @@ def version(args):
-                   '--add flag and --conn_uri flag: {invalid!r}\n')
-            msg = msg.format(invalid=invalid_args)
+        try:
+            new_conn = api_client.add_connection(
```
Might be easier to do `new_conn = api_client.add_connection(vars(args))`? (that does rely on the names from argparser matching the named args in the api client, but they do right now)

Not sure on this one.

ashb commented on airflow/bin/cli.py:
```diff
@@ -1110,133 +1108,77 @@ def version(args):
+                conn_extra=args.conn_extra)
+
+        except MissingArgument as ma:
+            print(ma)
```
Non-zero exit code here too

ashb commented on tests/api/common/experimental/connection_tests.py:
```diff
@@ -0,0 +1,427 @@
+            self.assertEqual(ia.exception, "Connections with the same id must all be of the same type")
+
+        # validate expected connections reached the DB
+        session = settings.Session()
```
Use `self.session`?

ashb commented on tests/api/common/experimental/connection_tests.py:
```diff
@@ -0,0 +1,427 @@
+                self.assertEqual(result, (conn_id, 'postgres', 'host', 5432,
+                                          extra[conn_id]))
+            elif conn_id == 'new4':
+                print('result: ', result)
```
Generally having bare prints in tests should be avoided - it gets noisy over time

ashb commented on tests/api/common/experimental/connection_tests.py:
```diff
@@ -0,0 +1,427 @@
+            password="airflow",
+            schema="airflow",
+            port="5432"
+        ))
```
From a quick look it appears that all the test cases use this connection - could we create it once in setUp() to reduce the amount of code in each test function?

ashb commented on tests/core.py:
```diff
@@ -1188,33 +1218,13 @@ def test_cli_connections_list(self):
-        # Check list attempt stdout
-        lines = [l for l in stdout.split('\n') if len(l) > 0]
-        self.assertListEqual(lines, [
-            ("\tThe following args are not compatible with the " +
```
What has happened to this test case? I think, it would still be useful, but I don't see it anymore?

caddac replied:
This test case was removed because the behavior of rejecting arguments for the list command was removed. It just lists the connections no matter what args are provided now.

ashb reviewed (changes requested):
A few small changes/comments. Looking good overall.

One thing that wasn't clear from a read of the code: does the API return the passwords on list or not? (I don't think it should, as right now none of the current methods of viewing connection in the CLI or WebUI give the value of the password)

ashb commented on tests/api/common/experimental/connection_tests.py:
```diff
@@ -0,0 +1,427 @@
+
+        session.query(models.Pool).delete()
+        session.query(models.Variable).delete()
+        session.query(models.Connection).delete()
```
This is causing your test failures:

>    AirflowException: The conn_id `sqlite_default` isn't defined

from elsewhere in the tests.

mik-laj commented on tests/core.py:
```diff
@@ -1248,75 +1258,166 @@ def test_cli_connections_add_delete(self):
+            "Successfully added `conn_id`=new3 : postgres://airflow:airflow@host:5432/airflow",
+            "Successfully added `conn_id`=new4 : postgres://airflow:airflow@host:5432/airflow",
+            "Successfully added `conn_id`=new5 : hive_metastore://airflow:airflow@host:9083/airflow",
+            "Successfully added `conn_id`=new6 : google_cloud_platform://:@:",
```
According to the "RFC 3986: Uniform Resource Identifier (URI): Generic Syntax" specification, this is not a valid address.
Schema part must be compatible with the format:
````
      scheme      = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
````
I created a issue https://issues.apache.org/jira/browse/AIRFLOW-3616  and working on PR: https://github.com/PolideaInternal/airflow/pull/30 
which deals with this topic.

mik-laj reviewed (changes requested):
Please improve compatibility with RFC 3986 Section 3.1

mik-laj commented on airflow/api/client/json_client.py:
```diff
@@ -89,3 +89,84 @@ def delete_pool(self, name):
+        :return: The updated Connection
+        """
+        endpoint = '/api/experimental/connections/'
+        url = urljoin(self._api_base_url, endpoint)
```
This should be done in the `_request` method. Can you correct this or will we do it in a separate PR?

mik-laj commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -253,3 +254,70 @@ def delete_pool(name):
+        _log.error(err)
+        response = jsonify(error="{}".format(err))
+        response.status_code = err.status_code
+        return response
```
This code snippet is repeated in any method. Should we think about how to avoid repeating this piece of code? Maybe introducing a decorator?
